packages/framework-core/src/booster-event-processor.ts

```diff
@@ -34,11 +35,32 @@ export class BoosterEventProcessor {
           BoosterEventProcessor.snapshotAndUpdateReadModels(config, entityName, entityID, eventStore, readModelStore)
         )
       }
+      // Store events that were just processed
+      await eventStore.storeProcessedEvents(unprocessedEvents)
 
       await Promises.allSettledAndFulfilled(eventEnvelopesProcessors)
     }
```

Comment:
@MarcAstr0 The operation of checking if the event and the insertion of that event in the table must be one atomic operation. If not, we could still process the event twice because the second event could come _after_ we searched for it but _before_ we store it into the processed table.

Also, we need to insert the event as soon as the event starts processing, not after it has been processed (for the same reason as above). I think it is a good idea to **change the name of the table to `dispatched_events`**. This way, it is clear that if an event is there, it has been dispatched, and it can be processed or being processed.

So the idea would be to only have one method in the `EventStore` called like `startEventProcessing` (the same method would be in each provider). That method will just try to insert the event in the `dispatched_events` table. If it succeeds, it returns true telling that you can proceed with the processing. If the insertion fails because of a "duplicate ID key" violation, it means the event was already dispatched, so the method will return false, meaning that event can't be processed (again).

This way, we achieve the insertion and "search for presence" operations in one atomic operation, avoiding any possible race condition when we have the two operations separated.

Reply:
@alvaroloes take a look at the latest changes, especially commit 7a26a48c3c3cf83d5733f6d11af36d974b8ea8c5. :wink: